lib/optimize/CommonsChunkPlugin.js

```diff
@@ -24,153 +24,304 @@ The available options are:
+		return {
+			chunkNames: chunkNames,
+			filenameTemplate: options.filename,
+			minChunks: options.minChunks,
+			selectedChunks: selectedChunks,
+			async: options.async,
+			minSize: options.minSize,
+			ident: __filename + (nextIdent++),
+		};
 	}
+
+	getCommonChunks(allChunks, compilation, chunkNames, selectedChunks, async) {
+		const asyncOrNoSelectedChunk = selectedChunks === false || async;
+
+		// we have specified chunk names
+		if(chunkNames) {
+			// map chunks by chunkName for quick access
+			const optimizedChunkMap = allChunks.reduce((map, chunk) => {
+				map.set(chunk.name, chunk);
+				return map;
+			}, new Map());
+
+			// Ensure we have a chunk per specified chunk name.
+			// Reuse existing chunks if possible
+			return chunkNames.map(chunkName => {
+				if(optimizedChunkMap.has(chunkName)) {
+					return optimizedChunkMap.get(chunkName);
+				}
+				// add the filtered chunks to the compilation
+				return compilation.addChunk(chunkName);
+			});
+		}
+
+		// we dont have named chunks specified, so we just take all of them
+		if(asyncOrNoSelectedChunk) {
+			return allChunks;
+		}
+
+		// that is not supposed to happen, lets throw
+		throw new Error("Invalid chunkNames argument");
+	}
+
+	getUsedChunks(compilation, allChunks, commonChunk, commonChunks, currentIndex, selectedChunks, isAsync) {
+		const asyncOrNoSelectedChunk = selectedChunks === false || isAsync;
+
+		if(Array.isArray(selectedChunks)) {
+			return allChunks.filter(chunk => {
+				const notCommmonChunk = chunk !== commonChunk;
+				const isSelectedChunk = selectedChunks.indexOf(chunk.name) > -1;
+				return notCommmonChunk && isSelectedChunk;
+			});
+		}
+
+		if(asyncOrNoSelectedChunk) {
+			// nothing to do here
+			if(!commonChunk.chunks) {
+				return [];
+			}
+
+			return commonChunk.chunks.filter((chunk) => {
+				// we can only move modules from this chunk if the "commonChunk" is the only parent
+				return isAsync || chunk.parents.length === 1;
+			});
+		}
+
+		// this is an entry point - bad
+		if(commonChunk.parents.length > 0) {
+			compilation.errors.push(new Error("CommonsChunkPlugin: While running in normal mode it's not allowed to use a non-entry chunk (" + commonChunk.name + ")"));
+			return;
+		}
+
+		// what is this?
+		return allChunks.filter((chunk) => {
+			const found = commonChunks.indexOf(chunk);
+			if(found >= currentIndex) return false;
+			return chunk.hasRuntime();
+		});
+	}
+
+	createAsyncChunk(compilation, asyncOption, commonChunk) {
+		const asyncChunk = compilation.addChunk(typeof asyncOption === "string" ? asyncOption : undefined);
+		asyncChunk.chunkReason = "async commons chunk";
+		asyncChunk.extraAsync = true;
+		asyncChunk.addParent(commonChunk);
+		return asyncChunk;
+	}
+
+	// If minChunks is a function use that
+	// otherwhise check if a module is used at least minChunks or 2 or usedChunks.length time
+	getModuleFilter(minChunks, commonChunk, usedChunksLength) {
```

Comment:
Instead of this: Always normalize the minChunks options to a function using the number and this code.